runtime/lua/vim/lsp/diagnostic.lua

```diff
@@ -298,15 +298,25 @@ function M.on_diagnostic(_, result, ctx, config)
   local uri = ctx.params.textDocument.uri
   local fname = vim.uri_to_fname(uri)
 
-  if result == nil then
-    return
-  end
-
-  if result.kind == 'unchanged' then
-    return
+  local diagnostics = {}
+  if result ~= nil then
+    if result.kind == 'unchanged' then
+      return
+    end
+    diagnostics = result.items
+  else
+    local client_name = vim.lsp.get_client_by_id(client_id).name
+    vim.notify_once(
+      string.format(
+        "LSP: client '%s' (id %s) returned nil to textDocument/diagnostic.\nThis is a bug in the LSP backend, and should be fixed within '%s'.",
+        client_name,
+        client_id,
+        client_name
+      ),
+      vim.log.levels.WARN
+    )
```

Comment:
While adding this ad-hoc warning message is fine in this PR, it should be noted that other handlers might throw errors (e.g. `workspace/applyEdit`) or silently ignore the invalid responses (e.g. `textDocument/definition`). (I'm not asking this should be fixed now) As we do further refactoring, this could be better handled in the common "boilerplate error validation and logging" wrapper for the handlers.